File: leet_2.py
class Node:
    def __init__(self, data):
        self.data = data 
        self.next = None 

class LinkedList:
    def __init__(self):
        self.head = None 
    
    def insert_new_node(self, data):
        node_n = Node(data) 

        if self.head is None:
            self.head = node_n 
            return 
        
        current = self.head 
        while current.next:
            current = current.next 
        current.next = node_n
        return self.head 
    
    # def add_two_number(self):
    #     val_one = ""
    #     val_two = 0 

    #     current = self.head 
    #     while current:
    #         val_one += str(current.data)
    #         current = current.next 

    #     return val_one[::-1]

    def __str__(self):
        res = [] 
        current = self.head 
        while current:
            res.append(current.data) 
            current = current.next 

        return str(res) 

l = LinkedList() 
l.insert_new_node(2) 
l.insert_new_node(4)
l.insert_new_node(3) 

m = LinkedList() 
m.insert_new_node(5)
m.insert_new_node(6) 
m.insert_new_node(4) 


class Soluation:
    def add_two_number(self, list_one : Node, list_two: Node):
        val_one = '' 
        current_one = list_one
        while current_one:
            val_one += str(current_one.data) 
            current_one = current_one.next 
        
        val_two = ''
        current_two = list_two 
        while current_two:
            val_two += str(current_two.data) 
            current_two = current_two.next 
        
        total = int(val_one[::-1]) + int(val_two[::-1]) 

        dummy = Node(0) 
        current = dummy 
        for digit in str(total)[::-1]:
            current.next = Node(int(digit)) 
            current = current.next 

        return dummy.next 

        
    

s = Soluation()
result = s.add_two_number(l.head, m.head)  
print(result)

        
        




# print(l) 

# print(m) 







# Input: l1 = [2,4,3], l2 = [5,6,4]
# Output: [7,0,8]
# Explanation: 342 + 465 = 807.

File: leet_2.ts
class Node {
  data: number;
  next: Node | null;

  constructor(data: number) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  head: Node | null;

  constructor() {
    this.head = null;
  }

  insertNewNode(data: number): void {
    const node = new Node(data);

    if (this.head === null) {
      this.head = node;
      return;
    }

    let current = this.head;
    while (current.next !== null) {
      current = current.next;
    }
    current.next = node;
  }

  toArray(): number[] {
    const result: number[] = [];
    let current = this.head;
    while (current !== null) {
      result.push(current.data);
      current = current.next;
    }
    return result;
  }
}

class Solution {
  addTwoNumbers(l1: Node | null, l2: Node | null): Node | null {
    let val1 = "";
    let val2 = "";

    let current1 = l1;
    let current2 = l2;

    while (current1 !== null) {
      val1 += current1.data.toString();
      current1 = current1.next;
    }

    while (current2 !== null) {
      val2 += current2.data.toString();
      current2 = current2.next;
    }

    const total =
      BigInt(val1.split("").reverse().join("")) +
      BigInt(val2.split("").reverse().join(""));
    const resultStr = total.toString().split("").reverse();

    const dummy = new Node(0);
    let current: Node = dummy;

    for (const digit of resultStr) {
      current.next = new Node(parseInt(digit));
      current = current.next;
    }

    return dummy.next;
  }
}


const l = new LinkedList();
l.insertNewNode(2);
l.insertNewNode(4);
l.insertNewNode(3);

const m = new LinkedList();
m.insertNewNode(5);
m.insertNewNode(6);
m.insertNewNode(4);

const s = new Solution();
let result = s.addTwoNumbers(l.head, m.head);

const resultArray: number[] = [];
while (result !== null) {
  resultArray.push(result.data);
  result = result.next;
}
console.log(resultArray);
